libmemcached/tls.c

```diff
@@ -0,0 +1,595 @@
+/**
+ * Copyright (C) 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#if defined(USE_TLS) && USE_TLS
+
+#include <openssl/err.h>
+#include <openssl/bio.h>
+#include <openssl/pem.h>
+#include <openssl/ssl.h>
+#include <openssl/opensslv.h>
+
+#include <libmemcached-1.0/tls.h>
+#include <libmemcached/common.h>
+#include <libmemcached/virtual_bucket.h>
+
+typedef enum {
+    WANT_READ = 1,
+    WANT_WRITE,
+    WANT_NONE
+} want_t;
+
+/* A wrapper for the openssl's SSL_CTX object. */
+typedef struct memcached_SSL_CTX {
+    /* Associated OpenSSL SSL_CTX as created by memcached_create_and_set_ssl_context() */
+    SSL_CTX *ctx;
+
+    /* Requested hostname for verefication.
+     * The hostname set needs to match with the CN (Common Name) of Subject of server certificate */
+    char *hostname;
+};
+
+/* A wrapper for the openssl's SSL object. */
+typedef struct memcached_SSL {
+    /* OpenSSL SSL object. */
+    SSL *ssl;
+
+    /* Store the default IO functions to switch back to in case TLS is disabled */
+    void *default_io_funcs;
+} memcached_SSL;
+
+int memcached_init_openssl(void)
+{
+    SSL_library_init();
+#ifdef HIREDIS_USE_CRYPTO_LOCKS
+    initOpensslLocks();
+#endif
+
+    return MEMCACHED_SUCCESS;
+}
+
+static context_funcs context_ssl_funcs = {
+    .free_privctx = memcached_free_ssl,
+    .read = memcached_ssl_read,
+    .write = memcached_ssl_write
+};
+
+void memcached_free_ssl_ctx(memcached_st* memc){
+    if (!memc) {
+        return;
+    }
+
+    _memcached_free_ssl_ctx(memc, (memcached_SSL_CTX*)memc->ssl_ctx);
+    memc->ssl_ctx = NULL;
+}
+
+void memcached_free_ssl(memcached_instance_st *instance){
+    if (instance == NULL) {
+        return;
+    }
+
+    memcached_SSL *memc_ssl =(memcached_SSL*)instance->privctx;
+    if (memc_ssl == NULL) {
+        return;
+    }
+    if (memc_ssl->ssl != NULL) {
+        SSL_free(memc_ssl->ssl);
+        memc_ssl->ssl = NULL;
+    }
+    instance->io_funcs = (context_funcs *)memc_ssl->default_io_funcs;
+    libmemcached_free(instance->root, memc_ssl);
+    instance->privctx = NULL;
+}
+
+/* Callback for passing a keyfile password stored as an sds to OpenSSL */
+static int ssl_set_password_callback(char *buf, int size, int rwflag, void *u) {
+    UNUSED(rwflag);
+
+    const char *pass = (const char *)u;
+    size_t pass_len;
+
+    if (!pass) return -1;
+    pass_len = strlen(pass);
+    if (pass_len > (size_t) size) return -1;
+    memcpy(buf, pass, pass_len);
+
+    return (int) pass_len;
+}
+
+/**
+ * Initiate a new SSL context
+ */
+static SSL_CTX* init_ctx(void)
+{
+    const SSL_METHOD *method;
+    SSL_CTX *ctx;
+
+    OpenSSL_add_all_algorithms();  /* Load cryptos, et.al. */
+    SSL_load_error_strings();   /* Bring in and register error messages */
+
+//#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+    method = TLS_client_method();
+//#else
+//    method = TLSv1_2_client_method();
+//#endif
```

Comment:
Can you remove the commented out code?

Reply:
It can be later used if we'll want to expend the use to openssl < 1.1.1

Reply:
removed